### DISCUSSION TITLE
add_filegroups没有达到预期效果

### DISCUSSION THREAD

**Discussion by Domain** [xmake-io/xmake#2407] [Q&A] [answered]
我用add_filegroups给vs工程里面的文件分组，但是好像无法达到想要的效果，我的目录结构是：
<pre>
--project
     |
     --- include
     |     |
     |     --- lib
     |          |
     |          ---common
     |                | 
     |                --- *.h
     --- src
     |     |
     |     --- lib
     |          |
     |          --- common
     |                 |
     |                 --- *.cpp / *.h
     |                 --- xmake.lua
</pre>
xmake.lua 内容如下：
```lua
target("common")
    set_kind("static")
    add_files("**.cpp")
    add_headerfiles("**.h")
    add_headerfiles("../../../include/lib/common/**.h")
    add_filegroups("include", { rootdir="../../../include/lib/common/", files={"**.h"}})
```
最终用vsxmake生成的工程中，仍然有..的目录：
<pre>
project
    |
    --- ..
    |    |
    |    ... include
</pre>
我看了一下vsxmake的lua代码，觉得是因为这句导致的：https://github.com/xmake-io/xmake/blob/50c5f96edb3f6c7a45e1bd24c921ec2c0afac33f/xmake/plugins/project/vsxmake/getinfo.lua#L364
这里fileitem是相对于rootdir的相对目录，而filepattern是绝对路径，match就不会成功，我改成filepath:match就正常了。

另外，其实我想连include这个都不要，直接合并两个目录，`add_filegroups("", { rootdir="../../../include/lib/common/", files={"**.h"}})`，但是这样会在路径前面加上"\\"，导致无法合并，我做了下面的修改，我还没有提PR，我不确定我理解是否有问题
```lua
                if filepath:match(filepattern) then
                    if mode == "plain" then
                        filter = path.normalize(filegroup)
                    else
                        -- file tree mode (default)
                        if filegroup == "" then
                            filter = path.normalize(path.directory(fileitem))
                        else
                            filter = path.normalize(path.join(filegroup, path.directory(fileitem)))
                        end
                    end
                    if filter and filter == '.' then
                        filter = nil
                    end
                    return filter  -- 已经找到了需要的filter，即使是nil也不应该break，否则没法合并根目录的文件
                end
```

**Comment by waruqi** [comment, marked as answer]
> 这里fileitem是相对于rootdir的相对目录，而filepattern是绝对路径，match就不会成功，我改成filepath:match就正常了。


你这个是 vsxmake generator 不是 vs generator。。


vs 工程之前改动后，相关 patch 同步到 vsxmake ，这个地方漏了一处同步，我刚更新了，再试下 dev

**Reply by Domain** [reply]
`add_filegroups("", { rootdir="../../../include/lib/common/", files={"**.h"}})`
这个写法考虑支持吗？

**Reply by waruqi** [reply]
必须要指定分组名，不能指定空串
